crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1114,6 +1293,111 @@ impl std::fmt::Display for LexicalErrorType {
+enum StringQuoteChar {
+    Single,
+    Double,
+}
+
+impl TryFrom<char> for StringQuoteChar {
+    type Error = String;
+
+    fn try_from(c: char) -> Result<Self, Self::Error> {
+        match c {
+            '\'' => Ok(Self::Single),
+            '"' => Ok(Self::Double),
+            _ => Err(format!("Invalid string quote character: {c}")),
+        }
+    }
+}
+
+impl StringQuoteChar {
+    const fn as_char(&self) -> char {
+        match self {
+            Self::Single => '\'',
+            Self::Double => '"',
+        }
+    }
+}
+
+#[derive(Debug)]
+enum StringQuoteSize {
+    Single,
+    Triple,
+}
+
+#[derive(Debug)]
+struct FStringContext {
+    quote_char: StringQuoteChar,
+    quote_size: StringQuoteSize,
+    parentheses_count: u32,
+    format_spec_count: u32,
+    is_raw_string: bool,
+}
+
+impl FStringContext {
+    fn new(quote_char: StringQuoteChar, quote_size: StringQuoteSize, is_raw_string: bool) -> Self {
+        Self {
+            quote_char,
+            quote_size,
+            parentheses_count: 0,
+            format_spec_count: 0,
+            is_raw_string,
+        }
+    }
+
+    /// Returns the quote character for the current f-string as a `char`.
+    const fn quote_char(&self) -> char {
+        self.quote_char.as_char()
+    }
+
+    /// Returns `true` if the current f-string allows multiline i.e., a triple-quoted f-string.
+    const fn allow_multiline(&self) -> bool {
+        matches!(self.quote_size, StringQuoteSize::Triple)
+    }
+
+    /// Returns `true` if the current f-string has open parentheses.
+    fn has_open_parentheses(&mut self) -> bool {
+        self.parentheses_count > 0
+    }
+
+    /// Increments the number of parentheses for the current f-string.
+    fn open_parentheses(&mut self) {
+        self.parentheses_count += 1;
+    }
+
+    /// Decrements the number of parentheses for the current f-string. If we're
+    /// in a format spec, also decrements the number of format specs.
+    fn close_parentheses(&mut self) {
+        if self.is_in_format_spec() {
+            self.format_spec_count = self.format_spec_count.saturating_sub(1);
+        }
+        self.parentheses_count = self.parentheses_count.saturating_sub(1);
+    }
+
+    /// Returns `true` if the lexer is in a f-string expression i.e., between two curly braces.
+    fn is_in_expression(&self) -> bool {
+        self.parentheses_count > self.format_spec_count
+    }
+
+    /// Returns `true` if the lexer is in a f-string format spec i.e., after a colon.
+    fn is_in_format_spec(&self) -> bool {
+        self.format_spec_count > 0 && !self.is_in_expression()
+    }
+
+    /// Returns `true` if the colon (`:`) for the current f-string is in a valid
+    /// position i.e., at the same level of nesting as the opening parentheses token.
+    /// Increments the number of format specs if it is.
+    fn is_valid_format_spec_position(&mut self) -> bool {
+        if self.parentheses_count - self.format_spec_count == 1 {
```

Comment:
This could potentially overflow, or is it guaranteed that `parentheses_count` is always larger than `format_spec_count` (even for invalid input)? 

Edit: This doesn't seem to be given, see `is_in_expression`. We need to use `saturating_sub` here